src/libsyntax/parse/parser.rs

```diff
@@ -3431,8 +3431,17 @@ impl<'a> Parser<'a> {
         let mut pats = Vec::new();
         loop {
             pats.push(self.parse_pat()?);
-            if self.check(&token::BinOp(token::Or)) { self.bump();}
-            else { return Ok(pats); }
+
+            if self.token == token::OrOr {
+                self.span_err_help(self.span,
+                                   "unexpected token `||` after pattern",
+                                   "did you mean to use `|` to specify multiple patterns?");
```

Comment:
When we have the relevant spans and know what code to suggest, the `.span_suggestion` method is arguably preferable (compact, attractive output; usable by tooling cc #42823).

```rust
let mut err = self.struct_span_err(self.span, "unexpected token `||` after pattern");
err.span_suggestion(self.span, "to specify multiple patterns, use `|`", "|".to_owned());
err.emit();
```

Reply:
I hope it's okay that I switched up the sentence a little bit because the exact text you have looks awkward when it's actually printed ` to specify multiple patterns use '|': '|'`